src/utils/api.ts

```diff
@@ -1,29 +1,67 @@
-import { Pokemon, PokemonData, PokemonDetails } from 'types';
+import { ErrorStatuses } from 'appConstants';
 
-export const fetchPokemons = async (url: string) => {
-  let pokemons: Array<Pokemon> = [];
+export const isServerError = (status: number) => String(status).startsWith(ErrorStatuses.server);
+
+export const hasError = (status: number) => !String(status).startsWith(ErrorStatuses.success);
+
+const handleCatch = (error: unknown) => {
+  error instanceof Error && window.alert(error.message);
+};
+
+export const handleResponse = async <T>(response: Response) => {
+  if (hasError(response.status)) {
+    if (isServerError(response.status)) {
+      window.alert('There was a server error');
+    }
+  } else return (await response.json()) as T;
+};
+
+export const handleMappedResponse = async <T>(arr: Array<Promise<Response>>) => {
+  const responses = await Promise.allSettled(arr);
+  const errors = responses.filter((res) => res.status === 'rejected').length;
+  const fullfilled = responses.filter((res) => res.status === 'fulfilled') as Array<
+    PromiseFulfilledResult<Response>
+  >;
+  const data = fullfilled.map((res) => res.value.json() as Promise<T>);
+  if (errors > 0) window.alert(`There was ${errors} errors on ${arr.length} requests.`);
+  return await Promise.all(data);
+};
+
+export const fetchPokemonByParameter = async <T>(url: string): Promise<T | null> => {
+  let results: T | null = null;
+  try {
+    const response = await fetch(url);
+    const data = await handleResponse<T>(response);
+    if (data) results = data;
+  } catch (error) {
+    handleCatch(error);
+  } finally {
+    return results;
+  }
+};
+
+export const fetchBase = async <T1 extends { results: Array<T2> }, T2>(url: string) => {
+  let base: Array<T2> = [];
   try {
-    const data = await fetch(url);
-    const { results }: PokemonData = await data.json();
-    pokemons = [...results];
+    const response = await fetch(url);
+    const data = await handleResponse<T1>(response);
+    if (data) base = [...data.results];
   } catch (error) {
-    console.error(error);
+    handleCatch(error);
   } finally {
-    return pokemons;
+    return base;
   }
 };
 
-export const fetchDetailedPokemons = async (pokemons: Array<Pokemon>) => {
-  let detailedPokemons: Array<PokemonDetails> = [];
+export const fetchDetails = async <T1 extends { url: string }, T2>(arr: Array<T1>) => {
+  let detailed: Array<T2> = [];
   try {
-    const responses = pokemons.map(async (pokemon) => {
-      const response = await fetch(pokemon.url);
-      return (await response.json()) as PokemonDetails;
-    });
-    detailedPokemons = await Promise.all(responses);
+    const responses = arr.map((item) => fetch(item.url).then((response) => response));
```

Comment:
I see potential issues with this code. And it is very hard readable and understandable. Why you are use this weird approach?

Reply:
When you write an application, constantly learning something new, sometimes a bit of a haos can sneak in. If I had to write everything from scratch, I would probably do a lot of things differently.

But I don't think it's very hard to read. The problem here was that If I want to have a set of information first I have to fetch available pokemons and then for each pokemon I need to fetch it's details. In this function I get array of pokemon objects [{ name, url}] and for each pokemon I have to fetch details using it's url. Then I'm handling returned responses

Reply:
Aftet all, I thought that it was unnecessarily overcomplicated and I changed the code